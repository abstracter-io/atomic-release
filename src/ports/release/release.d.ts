interface Release {
  getChangelog(): Promise<string | null>;

  getVersions(): Promise<string[]>;

  getNextVersion(): Promise<string>;

  getPreviousVersion(): Promise<string>;

  getMentionedIssues(): Promise<Set<string>>;

  getChangelogByVersion(version: string): Promise<string | null>;
}

export { Release };
