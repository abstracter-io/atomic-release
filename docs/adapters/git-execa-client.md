# GitExecaClient

An implemention of the SDK [GitClient](../ports/git-client.md) interface using [execa](https://github.com/sindresorhus/execa).

### Options

###### Optional properties are denoted by *

Type: `object literal`

##### remote*

Type: `string`  
Default: `origin`

##### workingDirectory*

Type: `string`  
Default: `process.cwd()`

### Example

```js
const { GitExecaClient } = require("@abstracter/atomic-release/adapters/git-execa-client");

const gitClient = new GitExecaClient({ remote: "origin2", workingDirectory: "/some/absolute/path" });
```
