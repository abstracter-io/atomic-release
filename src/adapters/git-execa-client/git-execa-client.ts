import execa from "execa";
import semver from "semver";
import { GitClient, MergedTag, Commit } from "../../ports/git-client";

type GitClientOptions = {
  remote: string;
  workingDirectory: string;
};

class GitExecaClient implements GitClient {
  private readonly options: GitClientOptions;

  public constructor(options?: Partial<GitClientOptions>) {
    this.options = {
      remote: options?.remote ?? "origin",
      workingDirectory: options?.workingDirectory ?? process.cwd(),
    };
  }

  private cli(args: string[], options?: execa.Options) {
    return execa("git", args, {
      cwd: this.options.workingDirectory,
      ...options,
    });
  }

  async log(range: string, format: string): Promise<string[]> {
    const logs: string[] = [];
    const delimiter = ":++:";
    const subprocess = await this.cli(["log", range, `--pretty=format:${format}${delimiter}`]);

    for (const log of subprocess.stdout.split(delimiter)) {
      if (log.length) {
        logs.push(log.startsWith("\n") ? log.substr(1) : log);
      }
    }

    return logs;
  }

  async cliVersion(): Promise<string> {
    const { stdout } = await this.cli(["--version"]);

    return stdout.split(" ")[2];
  }

  async refHash(ref: string): Promise<string> {
    const { stdout } = await this.cli(["rev-parse", ref]);

    return stdout;
  }

  async refName(ref: string): Promise<string> {
    const { stdout } = await this.cli(["rev-parse", "--abbrev-ref", ref]);

    return stdout;
  }

  /**
   * @param range A git log range.
   *
   * @example
   *
   * const gitClient = new GitClient();
   *
   * // Find all commits since ref 1234 (non-inclusive)
   * const commitSince = await gitClient.commits("1234..");
   *
   * // Find all commits until 1234 (inclusive)
   * const commitsUntil = await gitClient.commits("1234");
   */
  async commits(range: string): Promise<Commit[]> {
    const commits: Commit[] = [];
    const delimiter = ":<>:";
    const tagPattern = /tag: (.*),?/gi;
    // @src https://git-scm.com/docs/pretty-formats
    const formats = ["%H", "%s", "%b", "%N", "%D", "%ct", "%an", "%ae", "%cn", "%ce"];
    const rawCommits = await this.log(range, formats.join(delimiter));

    for (const rawCommit of rawCommits) {
      const tags: string[] = [];
      const logEntry = rawCommit.split(delimiter);

      let match: RegExpExecArray | null = null;

      for (const ref of logEntry[4].split(",")) {
        while ((match = tagPattern.exec(ref))) {
          tags.push(match[1]);
        }
      }

      commits.push({
        hash: logEntry[0],

        subject: logEntry[1],

        body: logEntry[2],

        notes: logEntry[3] ?? null,

        committedTimestamp: parseInt(logEntry[5], 10) * 1000,

        author: {
          name: logEntry[6],
          email: logEntry[7],
        },

        committer: {
          name: logEntry[8],
          email: logEntry[9],
        },

        tags,
      });
    }

    return commits;
  }

  async mergedTags(ref: string): Promise<MergedTag[]> {
    const minCliVersion = "2.7.0";
    const version = await this.cliVersion();

    // The following git command requires git CLI version to be >= 2.7.0
    // https://stackoverflow.com/a/39084124/1614199
    if (semver.satisfies(version, `>= ${minCliVersion}`)) {
      const tags: MergedTag[] = [];
      const delimiter = ":++:";
      const { stdout } = await this.cli([
        "tag",
        `--merged=${ref}`,
        `--format=%(refname:strip=2)${delimiter}%(objectname)`,
      ]);

      for (const tag of stdout.split("\n")) {
        if (tag.length) {
          const [name, hash] = tag.split(delimiter);

          tags.push({
            hash,
            name,
          });
        }
      }

      return tags;
    }
    //

    throw new Error(`Git version >= ${minCliVersion} is required. Found ${version}.`);
  }

  async remoteTagHash(tagName: string): Promise<string | null> {
    const { stdout } = await this.cli(["ls-remote", this.options.remote, "-t", `refs/tags/${tagName}`]);

    if (stdout.length) {
      return stdout.split("\t")[0];
    }

    return null;
  }

  async remoteBranchHash(branchName?: string): Promise<string | null> {
    const branch = branchName ?? (await this.refName("HEAD"));
    const { stdout } = await this.cli(["ls-remote", this.options.remote, "-h", `refs/heads/${branch}`]);

    if (stdout.length) {
      return stdout.split("\t")[0];
    }

    return null;
  }
}

export { GitExecaClient, GitClientOptions, MergedTag };
