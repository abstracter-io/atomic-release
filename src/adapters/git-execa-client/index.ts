export * from "./git-execa-client";
