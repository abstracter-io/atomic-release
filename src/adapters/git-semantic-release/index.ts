export * from "./git-semantic-release";
