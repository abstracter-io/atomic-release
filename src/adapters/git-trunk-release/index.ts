export * from "./git-trunk-release";
