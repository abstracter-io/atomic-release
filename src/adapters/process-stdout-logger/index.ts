export * from "./process-stdout-logger";
