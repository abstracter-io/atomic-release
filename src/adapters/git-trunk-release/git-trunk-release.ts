import getStream from "get-stream";
import intoStream from "into-stream";
import { ConventionalChangelogPreset } from "conventional-changelog-preset-loader";
import conventionalChangelogPreset from "conventional-changelog-conventionalcommits";
import conventionalChangelogWriter, { ChangelogWriterContext } from "conventional-changelog-writer";
import { ConventionalCommit, sync as syncConventionalCommitParser } from "conventional-commits-parser";

import { Logger } from "../../ports/logger";
import { memoize } from "../../utils/memoize";
import { Release } from "../../ports/release";
import { GitExecaClient } from "../git-execa-client";
import { processStdoutLogger } from "../process-stdout-logger";

type GitTrunkReleaseOptions = {
  logger?: Logger;

  gitClient?: GitExecaClient;

  remote?: string;

  workingDirectory?: string;

  changelogCommitFilter?: (commit: ConventionalCommit) => boolean;

  rawConventionalCommits?: (range: string) => Promise<{ hash: string; raw: string }[]>;

  conventionalChangelogPreset?: ConventionalChangelogPreset;

  conventionalChangelogWriterContext: ChangelogWriterContext | null;
};

type Options = Required<GitTrunkReleaseOptions>;

const defaultOptions = async (options: GitTrunkReleaseOptions): Promise<Options> => {
  const workingDirectory = options.workingDirectory ?? process.cwd();
  const remote = options.remote ?? "origin";
  const gitClient = options.gitClient ?? new GitExecaClient({
    remote,
    workingDirectory,
  });
  const rawConventionalCommits = async (range: string) => {
    const commits = await gitClient.commits(range);

    return commits.map((commit) => {
      const lines = [
        // subject
        `${commit.subject}`,

        // body
        `${commit.body}`,

        // extra fields
        "-hash-",
        `${commit.hash}`,

        "-gitTags-",
        `${commit.tags.join(",")}`,

        "-committerDate-",
        `${new Date(commit.committedTimestamp)}`,
      ];

      return {
        hash: commit.hash,
        raw: lines.join("\n"),
      };
    });
  };

  const changelogCommitFilter = (_commit: ConventionalCommit): boolean => {
    return true;
  };

  return {
    remote,
    gitClient,
    workingDirectory,

    logger: options.logger ?? processStdoutLogger({ name: "GitTrunkRelease" }),
    changelogCommitFilter: options.changelogCommitFilter ?? changelogCommitFilter,
    rawConventionalCommits: options.rawConventionalCommits ?? rawConventionalCommits,
    conventionalChangelogPreset: options.conventionalChangelogPreset ?? (await conventionalChangelogPreset()),
    conventionalChangelogWriterContext: options.conventionalChangelogWriterContext ?? null,
  };
};

const gitTrunkRelease = async (options: GitTrunkReleaseOptions): Promise<Release> => {
  const memo = memoize();
  const { logger, ...opt } = await defaultOptions(options);
  const preset = opt.conventionalChangelogPreset;
  const gitClient = opt.gitClient;
  const headHash = await gitClient.refHash("HEAD");

  // Private methods
  // ================
  const parseCommit = (rawConventionalCommit: string): ConventionalCommit => {
    return syncConventionalCommitParser(rawConventionalCommit, preset.parserOpts);
  };

  const getChangelogWriterContext = (): ChangelogWriterContext => {
    if (opt.conventionalChangelogWriterContext) {
      return opt.conventionalChangelogWriterContext;
    }

    throw new Error("conventional changelog writer context is missing");
  };

  const getConventionalCommits = async (): Promise<ConventionalCommit[]> => {
    return memo("conventional_commits", async () => {
      const commits = await opt.rawConventionalCommits("-1");
      const parsedCommits = commits.map((commit) => {
        return parseCommit(commit.raw);
      });

      return parsedCommits.filter(opt.changelogCommitFilter);
    });
  };

  const getPreviousVersionsConventionalCommits = async (): Promise<{ [version: string]: ConventionalCommit[] }> => {
    return memo("previous_versions_conventional_commits", async () => {
      const commits = await gitClient.commits("HEAD");
      const versionsConventionalCommits = {};

      // remove the first ('HEAD') commit
      commits.shift();

      await Promise.all(commits.map(async (commit) => {
        const [rawConventionalCommit] = await opt.rawConventionalCommits(`${commit.hash} -1`);
        const commits = [parseCommit(rawConventionalCommit.raw)];

        versionsConventionalCommits[commit.hash.slice(0, 7)] = commits.filter(opt.changelogCommitFilter);
      }));

      return versionsConventionalCommits;
    });
  };

  // Public methods
  // ==============
  const getVersions = async (): Promise<string[]> => {
    return memo("versions", async () => {
      const commits = await gitClient.commits("HEAD");

      commits.shift();

      return commits.map((commit) => {
        return commit.hash.slice(0, 7);
      });
    });
  };

  const getChangelog = async (): Promise<string | null> => {
    return memo("changelog", async () => {
      const nextVersion = await getNextVersion();
      const commits = await getConventionalCommits();

      if (commits.length) {
        const writerOpts = preset.writerOpts;
        const writerContext = {
          ...getChangelogWriterContext(),
          version: nextVersion,
        };
        const stream = intoStream.object(commits).pipe(conventionalChangelogWriter(writerContext, writerOpts));

        return getStream(stream);
      }

      return null;
    });
  };

  const getNextVersion = async (): Promise<string> => {
    return memo("next_version", async () => {
      return headHash.slice(0, 7);
    });
  };

  const getPreviousVersion = async (): Promise<string> => {
    return memo("previous_version", async () => {
      const versions = await getVersions();

      if (!versions.length) {
        const hash = headHash.slice(0, 7);

        logger.info(`Could not find a previous version. Will use ${hash} as initial version`);

        return hash;
      }

      return versions[0];
    });
  };

  const getMentionedIssues = async (): Promise<Set<string>> => {
    return memo("mentioned_issues", async () => {
      const issues = new Set<string>();
      const commits = await getConventionalCommits();

      for (const commit of commits) {
        for (const reference of commit.references) {
          const issue = reference.issue;

          if (issue) {
            issues.add(issue);
          }
        }
      }

      return issues;
    });
  };

  const getChangelogByVersion = async (version: string): Promise<string | null> => {
    return memo(`changelog_${version}`, async () => {
      const versionsConventionalCommits = await getPreviousVersionsConventionalCommits();
      const commits = versionsConventionalCommits[version];

      if (Array.isArray(commits)) {
        if (commits.length) {
          const writerOpts = preset.writerOpts;
          const writerContext = {
            ...getChangelogWriterContext(),
            version,
          };
          const stream = intoStream.object(commits).pipe(conventionalChangelogWriter(writerContext, writerOpts));

          return getStream(stream);
        }

        return null;
      }

      throw new Error(`Could not find commits for version '${version}'`);
    });
  };

  return {
    getVersions,
    getChangelog,
    getNextVersion,
    getMentionedIssues,
    getPreviousVersion,
    getChangelogByVersion,
  };
};

export { gitTrunkRelease, GitTrunkReleaseOptions };
