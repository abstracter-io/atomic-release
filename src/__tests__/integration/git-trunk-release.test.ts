import { Release } from "../../ports/release";
import { GitExecaClient } from "../../adapters/git-execa-client";
import { gitTrunkRelease } from "../../adapters/git-trunk-release";

import { spiedLogger } from "../stubs/spied-logger";

const LOGGER = spiedLogger();
const HASH = "c658ea3e060490dced90dfb34c018d88b8e797f9";

class GitClientStub extends GitExecaClient {
  constructor() {
    super({ workingDirectory: process.cwd() });
  }

  cliVersion = jest.fn();
  refHash = jest.fn();
  refName = jest.fn();
  commits = jest.fn();
  mergedTags = jest.fn();
  remoteTagHash = jest.fn();
  remoteBranchHash = jest.fn();
}

const releaseOptions = () => {
  return {
    logger: LOGGER,
    conventionalChangelogWriterContext: {
      owner: "t",
      repository: "t",
      host: "https://github.com",
      repoUrl: "https://github.com/t/t",
      date: "2021-10-08",
    },
  };
};

const commit = () => {
  return {
    subject: "feat: ...",
    body: "",
    notes: "",
    author: {
      name: "",
      email: "",
    },
    committer: {
      name: "",
      email: "",
    },
    tags: [],
    hash: HASH,
    committedTimestamp: 1633686020134,
  };
};

describe("git trunk release", () => {
  let release: Release;
  let gitClient: jest.Mocked<GitExecaClient>;

  beforeEach(async () => {
    gitClient = new GitClientStub() as unknown as jest.Mocked<GitExecaClient>;

    gitClient.refHash.mockImplementation(async () => {
      return HASH;
    });

    gitClient.commits.mockImplementation(async () => [commit()]);

    gitClient.refName.mockImplementation(async () => "main");

    release = await gitTrunkRelease({
      gitClient,
      ...releaseOptions(),
    });
  });

  test("changelog is null", async () => {
    const changelogCommitFilter = jest.fn();
    const release = await gitTrunkRelease({
      ...releaseOptions(),
      gitClient,
      changelogCommitFilter,
    });

    expect(await release.getChangelog()).toBeNull();
    expect(changelogCommitFilter).toBeCalledTimes(1);
  });

  test("changelog is generated", async () => {
    const release = await gitTrunkRelease({
      ...releaseOptions(),

      gitClient,

      rawConventionalCommits: async (range: string) => {
        const commits = await gitClient.commits(range);

        return commits.map((commit) => {
          const lines = [
            // subject
            `${commit.subject}`,

            // body
            `${commit.body}`,

            // extra fields
            "-hash-",
            `${commit.hash}`,

            "-gitTags-",
            `${commit.tags.join(",")}`,

            "-committerDate-",
            `${new Date(1633686020134)}`,
          ];

          return {
            hash: commit.hash,
            raw: lines.join("\n"),
          };
        });
      },
    });

    const changelog = await release.getChangelog();

    expect(changelog).toMatchSnapshot();
  });

  test("next version is 'HEAD' hash", async () => {
    const expectedHash = HASH.slice(0, 7);

    expect(await release.getNextVersion()).toStrictEqual(expectedHash);
  });

  test("list issues mentioned in commits", async () => {
    gitClient.commits.mockImplementation(async () => {
      return [
        {
          ...commit(),
          subject: "feat!: ... closes #3, #46, #39",
        },
        { ...commit(), subject: "Merge pull request #999 from repo/branch" },
      ];
    });

    expect(await release.getMentionedIssues()).toStrictEqual(new Set(["3", "39", "46", "999"]));
  });

  test("previous version changelog is null", async () => {
    const hash = "10f03409c73ffa37fbd2b890d99c74c63d0f9f03";
    const changelogCommitFilter = jest.fn(() => false);
    const release = await gitTrunkRelease({
      ...releaseOptions(),

      gitClient,

      changelogCommitFilter,
    });

    gitClient.commits.mockImplementation(async () => {
      return [commit(), { ...commit(), hash }];
    });

    expect(await release.getChangelogByVersion(hash.slice(0, 7))).toBeNull();
    expect(changelogCommitFilter).toBeCalledTimes(1);
  });

  test("previous version changelog is generated", async () => {
    const hash = "10f03409c73ffa37fbd2b890d99c74c63d0f9f03";

    gitClient.commits.mockImplementation(async () => {
      return [commit(), { ...commit(), hash }];
    });

    const changelog = await release.getChangelogByVersion(hash.slice(0, 7));

    expect(changelog).not.toBeNull();
    expect(changelog).toMatchSnapshot();
  });

  test("previous version fallbacks to 'HEAD' hash", async () => {
    const hash = HASH.slice(0, 7);

    expect(await release.getPreviousVersion()).toStrictEqual(hash);

    expect(LOGGER.info).toBeCalledWith(`Could not find a previous version. Will use ${hash} as initial version`);
  });

  test("previous version is the second commit hash", async () => {
    const expectedVersion = HASH.slice(0, 7);

    gitClient.commits.mockImplementation(async () => {
      return [
        { ...commit(), hash: "1234" },

        { ...commit(), hash: HASH },
      ];
    });

    expect(await release.getPreviousVersion()).toStrictEqual(expectedVersion);
  });

  test("rawConventionalCommits is called with 'HEAD' ref", async () => {
    const rawConventionalCommits = jest.fn();
    const release = await gitTrunkRelease({
      ...releaseOptions(),

      gitClient,

      rawConventionalCommits,
    });

    rawConventionalCommits.mockImplementation(() => []);

    await release.getChangelog();

    expect(rawConventionalCommits).toBeCalledWith(`HEAD -1`);
  });

  test("generating changelog fails when writer context is missing", async () => {
    const release = await gitTrunkRelease({
      ...releaseOptions(),
      gitClient,
      conventionalChangelogWriterContext: null as never,
    });
    const expectedError = new Error(`conventional changelog writer context is missing`);

    expect(await release.getChangelog().catch((e) => e)).toStrictEqual(expectedError);
  });

  test("generating changelog by version fails when version does not exists", async () => {
    const version = "fffferrt";
    const expectedError = new Error(`Could not find commits for version '${version}'`);

    expect(await release.getChangelogByVersion(version).catch((e) => e)).toStrictEqual(expectedError);
  });
});
