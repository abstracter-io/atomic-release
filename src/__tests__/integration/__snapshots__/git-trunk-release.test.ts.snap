// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`git trunk release changelog is generated 1`] = `
"## c658ea3 (2021-10-08)


### Features

* ... ([c658ea3](https://github.com/t/t/commit/c658ea3e060490dced90dfb34c018d88b8e797f9))

"
`;

exports[`git trunk release previous version changelog is generated 1`] = `
"## 10f0340 (2021-10-08)


### Features

* ... ([c658ea3](https://github.com/t/t/commit/c658ea3e060490dced90dfb34c018d88b8e797f9))

"
`;
